plugins/database/redshift/redshift.go

```diff
@@ -229,132 +172,180 @@ func (r *RedShift) CreateUser(ctx context.Context, statements dbplugin.Statement
-				return err
+	if req.Password != nil {
+		username := req.Username
+		password := req.Password.NewPassword
+		if username == "" || password == "" {
+			return dbplugin.UpdateUserResponse{}, errors.New("must provide both username and password")
+		}
+
+		// Check if the role exists
+		var exists bool
+		err = db.QueryRowContext(ctx, "SELECT exists (SELECT usename FROM pg_user WHERE usename=$1);", username).Scan(&exists)
+		if err != nil && err != sql.ErrNoRows {
+			return dbplugin.UpdateUserResponse{}, err
+		}
+
+		// Vault requires the database user already exist, and that the credentials
+		// used to execute the rotation statements has sufficient privileges.
+		statements := req.Password.Statements.Commands
+		if len(statements) == 0 {
+			statements = []string{defaultRotateRootCredentialsSQL}
+		}
+		// Execute each query
+		for _, stmt := range statements {
+			for _, query := range strutil.ParseArbitraryStringSlice(stmt, ";") {
+				query = strings.TrimSpace(query)
+				if len(query) == 0 {
+					continue
+				}
+
+				m := map[string]string{
+					"name":     username,
+					"username": username,
+					"password": password,
+				}
+				if err := dbtxn.ExecuteTxQuery(ctx, tx, m, query); err != nil {
+					return dbplugin.UpdateUserResponse{}, err
+				}
 			}
 		}
 	}
 
-	return tx.Commit()
+	return dbplugin.UpdateUserResponse{}, tx.Commit()
 }
 
-func (r *RedShift) RevokeUser(ctx context.Context, statements dbplugin.Statements, username string) error {
+// DeleteUser supports both default and custom statements to delete a user.
+func (r *RedShift) DeleteUser(ctx context.Context, req dbplugin.DeleteUserRequest) (dbplugin.DeleteUserResponse, error) {
 	// Grab the lock
 	r.Lock()
 	defer r.Unlock()
 
-	statements = dbutil.StatementCompatibilityHelper(statements)
-
-	if len(statements.Revocation) == 0 {
-		return r.defaultRevokeUser(ctx, username)
+	if len(req.Statements.Commands) == 0 {
+		return r.defaultDeleteUser(ctx, req)
 	}
 
-	return r.customRevokeUser(ctx, username, statements.Revocation)
+	return r.customDeleteUser(ctx, req)
 }
 
-func (r *RedShift) customRevokeUser(ctx context.Context, username string, revocationStmts []string) error {
+func (r *RedShift) customDeleteUser(ctx context.Context, req dbplugin.DeleteUserRequest) (dbplugin.DeleteUserResponse, error) {
 	db, err := r.getConnection(ctx)
 	if err != nil {
-		return err
+		return dbplugin.DeleteUserResponse{}, err
 	}
 	defer db.Close()
 
 	tx, err := db.BeginTx(ctx, nil)
 	if err != nil {
-		return err
+		return dbplugin.DeleteUserResponse{}, err
 	}
 	defer func() {
 		tx.Rollback()
 	}()
 
-	for _, stmt := range revocationStmts {
+	for _, stmt := range req.Statements.Commands {
 		for _, query := range strutil.ParseArbitraryStringSlice(stmt, ";") {
 			query = strings.TrimSpace(query)
 			if len(query) == 0 {
 				continue
 			}
 
 			m := map[string]string{
-				"name": username,
+				"name": req.Username,
```

Comment:
Should this also have `"username"` as an option? That way we'd be consistent.

Reply:
`username` was only used in the default root rotation statement, and wasn't documented anywhere as a supported template. However, I do agree consistency within the `UpdateUser` function at least would be nice. Do you think we should spread the consistency further to include `{{username}}` for `NewUser` and `DeleteUser`?

Reply:
At this point I think we should. It's been a mix of `user` and `username` across databases (and even across specific DB calls) so I think just making it available as both everywhere is the right call.